components/core/src/clp/hash_utils.hpp

```diff
@@ -0,0 +1,122 @@
+ */
+class EvpDigestContext {
+public:
+    // Types
+    class OperationFailed : public TraceableException {
+    public:
+        // Constructors
+        OperationFailed(ErrorCode error_code, char const* const filename, int line_number)
+                : TraceableException(error_code, filename, line_number) {}
+
+        // Methods
+        [[nodiscard]] auto what() const noexcept -> char const* override {
+            return "EvpCtxManager operation failed";
+        }
+    };
+
+    // Constructors
+    /**
+     * Creates and initializes a hash context by calling EVP_MD_CTX_create
+     * and EVP_DigestInit_ex from OpenSSL library
+     * @param type the type of digest (hash algorithm) from OpenSSL library.
+     * @throw EvpCtxManager::ErrorCode_Failure if EVP_MD_CTX_create fails.
+     * @throw EvpCtxManager::ErrorCode_NotInit if EVP_DigestInit_ex fails.
+     */
+    EvpDigestContext(EVP_MD const* type) : m_md_ctx{EVP_MD_CTX_create()} {
+        if (nullptr == m_md_ctx) {
+            throw OperationFailed(ErrorCode_Failure, __FILENAME__, __LINE__);
+        }
+        // Set impl to nullptr to use the default implementation of digest type
+        if (1 != EVP_DigestInit_ex(m_md_ctx, type, nullptr)) {
+            throw OperationFailed(ErrorCode_NotInit, __FILENAME__, __LINE__);
+        }
+    }
+
+    // Disable copy constructor/assignment operator
+    EvpDigestContext(EvpDigestContext const&) = delete;
+    auto operator=(EvpDigestContext const&) -> EvpDigestContext& = delete;
+
+    // disable move constructor/assignment operator
+    EvpDigestContext(EvpDigestContext&&) = delete;
+    auto operator=(EvpDigestContext&&) -> EvpDigestContext& = delete;
+
+    // Destructor
+    ~EvpDigestContext() { EVP_MD_CTX_destroy(m_md_ctx); }
+
+    // Methods
+    /**
+     * Calls EVP_DigestUpdate to digest the input and update the hash context
+     * @param input
+     * @return ErrorCode_Success on success.
+     * @return ErrorCode_Unsupported if context is already finalized.
+     * @return ErrorCode_Failure if EVP_DigestUpdate fails.
+     */
+    [[nodiscard]] auto digest_update(std::span<unsigned char const> input) -> ErrorCode;
+
+    /**
+     * Calls EVP_DigestFinal_ex to generate the final hash result
+     * @param hash Returns the hashing result.
+     * @return ErrorCode_Success on success.
+     * @return ErrorCode_Unsupported if context is already finalized.
+     * @return ErrorCode_Corrupt if the hashing result has an unexpected length.
+     * @return ErrorCode_Failure if EVP_DigestFinal_ex fails.
+     */
+    [[nodiscard]] auto digest_final_ex(std::vector<unsigned char>& hash) -> ErrorCode;
+
+private:
+    EVP_MD_CTX* m_md_ctx{nullptr};
+    bool m_is_digest_finalized{false};
+};
+
+/**
+ * Converts input hash into a hex string.
+ * @param input
+ * @return input hash as a hex string
+ */
+[[nodiscard]] auto convert_hash_to_hex_string(std::span<unsigned char> input) -> std::string;
```

Comment:
How about `convert_to_hex_string` since this method is not really specific to hashes.